packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -75,9 +80,48 @@ module.exports = class RequestClient {
     return res.json()
   }
 
+  preflight (path) {
+    return new Promise((resolve, reject) => {
+      if (this.preflightDone) {
+        return resolve(this.allowedHeaders.slice())
+      }
+
+      fetch(this._getUrl(path), {
+        method: 'OPTIONS'
+      })
+        .then((response) => {
+          if (response.headers.has('access-control-allow-headers')) {
+            this.allowedHeaders = response.headers.get('access-control-allow-headers')
+              .split(',').map((headerName) => headerName.trim().toLowerCase())
+          }
+          this.preflightDone = true
+          resolve(this.allowedHeaders.slice())
+        })
+        .catch(reject)
+    })
+  }
+
+  preflightAndHeaders (path) {
+    return new Promise((resolve, reject) => {
```

Comment:
This promise chain would really benefit from being flatter I think…

Right now the `this.headers()` call does not have a catch() handler.

Reply:
changed 👍